
<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">


# Logistic Regresion Lab
## Exercise with bank marketing data

_Authors: Sam Stack(DC)_

## Introduction
- Data from the UCI Machine Learning Repository: data, [data dictionary](https://archive.ics.uci.edu/ml/datasets/Bank+Marketing)
- **Goal**: Predict whether a customer will purchase a bank product marketed over the phone
- `bank-additional.csv` is already in our repo, so there is no need to download the data from the UCI website

## Step 1: Read the data into Pandas

In [69]:
import pandas as pd
import matplotlib.pyplot as plt
bank = pd.read_csv('../datasets/bank.csv')
bank.head()

age          job  marital          education default  housing     loan  \
0   30  blue-collar  married           basic.9y      no      yes       no   
1   39     services   single        high.school      no       no       no   
2   25     services  married        high.school      no      yes       no   
3   38     services  married           basic.9y      no  unknown  unknown   
4   47       admin.  married  university.degree      no      yes       no   

     contact month day_of_week ...  campaign  pdays  previous     poutcome  \
0   cellular   may         fri ...         2    999         0  nonexistent   
1  telephone   may         fri ...         4    999         0  nonexistent   
2  telephone   jun         wed ...         1    999         0  nonexistent   
3  telephone   jun         fri ...         3    999         0  nonexistent   
4   cellular   nov         mon ...         1    999         0  nonexistent   

  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed  y  
0         -1.8          92.893          -46.2      1.313       5099.1  0  
1          1.1          93.994          -36.4      4.855       5191.0  0  
2          1.4          94.465          -41.8      4.962       5228.1  0  
3          1.4          94.465          -41.8      4.959       5228.1  0  
4         -0.1          93.200          -42.0      4.191       5195.8  0  

[5 rows x 21 columns]

**  Target '`y`' represented as such**
    - No : 0
    - Yes : 1
    


In [70]:
bank.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

Notes:
- Data set is imbalanced, most people (89%) do not purchase
- Don't know duration before a call - can't be used to predict outcome
- pdays variable is 999 96% of the time - perhaps not useful

In [71]:
# Perform what ever steps you need to familiarize yourself with the data:
bank.describe()

age     duration     campaign        pdays     previous  \
count  4119.000000  4119.000000  4119.000000  4119.000000  4119.000000   
mean     40.113620   256.788055     2.537266   960.422190     0.190337   
std      10.313362   254.703736     2.568159   191.922786     0.541788   
min      18.000000     0.000000     1.000000     0.000000     0.000000   
25%      32.000000   103.000000     1.000000   999.000000     0.000000   
50%      38.000000   181.000000     2.000000   999.000000     0.000000   
75%      47.000000   317.000000     3.000000   999.000000     0.000000   
max      88.000000  3643.000000    35.000000   999.000000     6.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx    euribor3m  nr.employed  \
count   4119.000000     4119.000000    4119.000000  4119.000000  4119.000000   
mean       0.084972       93.579704     -40.499102     3.621356  5166.481695   
std        1.563114        0.579349       4.594578     1.733591    73.667904   
min       -3.400000       92.201000     -50.800000     0.635000  4963.600000   
25%       -1.800000       93.075000     -42.700000     1.334000  5099.100000   
50%        1.100000       93.749000     -41.800000     4.857000  5191.000000   
75%        1.400000       93.994000     -36.400000     4.961000  5228.100000   
max        1.400000       94.767000     -26.900000     5.045000  5228.100000   

                 y  
count  4119.000000  
mean      0.109493  
std       0.312294  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000

In [72]:
# what does the target look like? imbalanced
bank.y.value_counts()/len(bank)

0    0.890507
1    0.109493
Name: y, dtype: float64

In [73]:
bank.day_of_week.value_counts()/len(bank)

thu    0.208789
mon    0.207575
tue    0.204176
wed    0.193008
fri    0.186453
Name: day_of_week, dtype: float64


## Step 2: Prepare at least three features
- Include both numeric and categorical features
- Choose features that you think might be related to the response (based on intuition or exploration)
- Think about how to handle missing values (encoded as "unknown")

In [132]:
# A: age, euribor3m, campaign, default
# add day_of_week feature with dummies
#data = bank.join(pd.get_dummies(bank.day_of_week, prefix='bank_', drop_first=True))
#data.columns()


## Step 3: Model building
- Use cross-validation to evaluate the logistic regression model with your chosen features.  
    You can use any (combination) of the following metrics to evaluate.
    - [Classification/Accuracy Error](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html)
    - [Confusion Matrix](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html)
    - [ROC curves and area under a curve (AUC)](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score)
    - [Log loss](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html)
- Try to increase the metrics by selecting different sets of features
    - *Bonus*: Experiment with hyper parameters such are regularization.

In [36]:
from sklearn.linear_model import LogisticRegression

from sklearn.model_selection import train_test_split

from sklearn import metrics

**Build a Model**  

In [106]:
bank_a = bank[bank['job'] != 'unknown']
bank_a = bank_a[bank_a['education'] != 'unknown']
bank_a.head()

age          job  marital          education default  housing     loan  \
0   30  blue-collar  married           basic.9y      no      yes       no   
1   39     services   single        high.school      no       no       no   
2   25     services  married        high.school      no      yes       no   
3   38     services  married           basic.9y      no  unknown  unknown   
4   47       admin.  married  university.degree      no      yes       no   

     contact month day_of_week ...  campaign  pdays  previous     poutcome  \
0   cellular   may         fri ...         2    999         0  nonexistent   
1  telephone   may         fri ...         4    999         0  nonexistent   
2  telephone   jun         wed ...         1    999         0  nonexistent   
3  telephone   jun         fri ...         3    999         0  nonexistent   
4   cellular   nov         mon ...         1    999         0  nonexistent   

  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed  y  
0         -1.8          92.893          -46.2      1.313       5099.1  0  
1          1.1          93.994          -36.4      4.855       5191.0  0  
2          1.4          94.465          -41.8      4.962       5228.1  0  
3          1.4          94.465          -41.8      4.959       5228.1  0  
4         -0.1          93.200          -42.0      4.191       5195.8  0  

[5 rows x 21 columns]

In [125]:
test = list(data.columns).remove('age')
test # why doesn't this work?

In [126]:
# convert selected features do dummies
data = pd.get_dummies(bank_a[['age','job','education','day_of_week','y']], drop_first=True)

# set the model
logreg = LogisticRegression()

# set x and y
feature_cols = ['age', 'job_blue-collar', 'job_entrepreneur', 'job_housemaid',
       'job_management', 'job_retired', 'job_self-employed', 'job_services',
       'job_student', 'job_technician', 'job_unemployed', 'education_basic.6y',
       'education_basic.9y', 'education_high.school', 'education_illiterate',
       'education_professional.course', 'education_university.degree',
       'day_of_week_mon', 'day_of_week_thu', 'day_of_week_tue',
       'day_of_week_wed']
X = data[feature_cols]
y = data.y

# train test splot
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=123)

# fit model
logreg.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

**Get the Coefficient for each feature.**
- Be sure to make note of interesting findings.



**Use the Model to predict on x_test and evaluate the model using metric(s) of Choice.**

In [127]:
# A:
y_preds = logreg.predict(X_test)

# check performance
metrics.confusion_matrix(y_test, y_preds)

array([[866,   0],
       [116,   0]])

In [128]:
print('Accuracy: {:.3f}'.format(914/(914+116)))
print('Recall / true positive rate: {}'.format(0))

Accuracy: 0.887
Recall / true positive rate: 0


In [129]:
metrics.accuracy_score(y_test, y_preds)

0.88187372708757639

In [130]:
# baseline accuracy
[bank.y.value_counts()/len(bank)][0][0]

0.89050740470988099

In [131]:
# roc auc
metrics.roc_auc_score(y_test, y_preds)

0.5

### Model 2: Use a different combination of features.
- Evaluate the model and interpret your choosen metrics.

In [ ]:
# A;



### Is your model not performing very well?

Is it not predicting any True Positives?

Lets try one more thing before we revert to grabbing more features.  Adjusting the probability threshold.

Use the `LogisticRegression.predict_proba()` attribute to get the probabilities.

Recall from the lesson the first probability is the for `class 0` and the second is for `class 1`.

In [ ]:
# A:

**Visualize the distribution**

In [ ]:
# A:

** Calculate a new threshold and use it to convert predicted probabilities to output classes**



In [ ]:
# A:

**Evaluate the model metrics now**

In [ ]:
# A:

## Step 4: Build a model using all of the features.

- Evaluate it using your prefered metrics.

In [ ]:
# A:

## Bonus: Use Regularization to optimize your model.

In [ ]:
# try using a for loop to test various regularization strengths 'C'